epochStart/bootstrap/simpleEpochStartMetaBlockInterceptor.go

```diff
@@ -0,0 +1,131 @@
+package bootstrap
+
+import (
+	"math"
+	"sync"
+	"time"
+
+	"github.com/ElrondNetwork/elrond-go/core"
+	"github.com/ElrondNetwork/elrond-go/core/check"
+	"github.com/ElrondNetwork/elrond-go/data/block"
+	"github.com/ElrondNetwork/elrond-go/hashing"
+	"github.com/ElrondNetwork/elrond-go/marshal"
+	"github.com/ElrondNetwork/elrond-go/p2p"
+)
+
+const timeToWaitBeforeCheckingReceivedHeaders = 1 * time.Second
+const numTriesUntilExit = 5
+
+type simpleEpochStartMetaBlockInterceptor struct {
+	marshalizer            marshal.Marshalizer
+	hasher                 hashing.Hasher
+	mutReceivedMetaBlocks  sync.RWMutex
+	mapReceivedMetaBlocks  map[string]*block.MetaBlock
+	mapMetaBlocksFromPeers map[string][]p2p.PeerID
+}
+
+// NewSimpleEpochStartMetaBlockInterceptor will return a new instance of simpleEpochStartMetaBlockInterceptor
+func NewSimpleEpochStartMetaBlockInterceptor(marshalizer marshal.Marshalizer, hasher hashing.Hasher) (*simpleEpochStartMetaBlockInterceptor, error) {
+	if check.IfNil(marshalizer) {
+		return nil, ErrNilMarshalizer
+	}
+	if check.IfNil(hasher) {
+		return nil, ErrNilHasher
+	}
+
+	return &simpleEpochStartMetaBlockInterceptor{
+		marshalizer:            marshalizer,
+		hasher:                 hasher,
+		mutReceivedMetaBlocks:  sync.RWMutex{},
+		mapReceivedMetaBlocks:  make(map[string]*block.MetaBlock),
+		mapMetaBlocksFromPeers: make(map[string][]p2p.PeerID),
+	}, nil
+}
+
+// ProcessReceivedMessage will receive the metablocks and will add them to the maps
+func (s *simpleEpochStartMetaBlockInterceptor) ProcessReceivedMessage(message p2p.MessageP2P, _ func(buffToSend []byte)) error {
+	var mb block.MetaBlock
+	err := s.marshalizer.Unmarshal(&mb, message.Data())
+	if err != nil {
+		return err
+	}
+	s.mutReceivedMetaBlocks.Lock()
+	mbHash, err := core.CalculateHash(s.marshalizer, s.hasher, &mb)
+	if err != nil {
+		s.mutReceivedMetaBlocks.Unlock()
+		return err
+	}
+	s.mapReceivedMetaBlocks[string(mbHash)] = &mb
+	s.addToPeerList(string(mbHash), message.Peer())
+	s.mutReceivedMetaBlocks.Unlock()
+
+	return nil
+}
+
+// this func should be called under mutex protection
+func (s *simpleEpochStartMetaBlockInterceptor) addToPeerList(hash string, id p2p.PeerID) {
+	peersListForHash, ok := s.mapMetaBlocksFromPeers[hash]
+
+	if !ok {
+		s.mapMetaBlocksFromPeers[hash] = append(s.mapMetaBlocksFromPeers[hash], id)
+		return
+	}
+
+	for _, peer := range peersListForHash {
+		if peer == id {
+			return
+		}
+	}
+
+	s.mapMetaBlocksFromPeers[hash] = append(s.mapMetaBlocksFromPeers[hash], id)
+}
+
+// GetEpochStartMetaBlock will return the metablock after it is confirmed or an error if the number of tries was exceeded
+func (s *simpleEpochStartMetaBlockInterceptor) GetEpochStartMetaBlock(target int, epoch uint32) (*block.MetaBlock, error) {
+	for count := 0; count < numTriesUntilExit; count++ {
+		time.Sleep(timeToWaitBeforeCheckingReceivedHeaders)
```

Comment:
it would be better to do with a channel, but could stay for the next PR, add TODO

Reply:
done